core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -9633,6 +9641,32 @@ private boolean isUserStoreChainNeeded(List<String> userStorePreferenceOrder) th
     private boolean generateUserStoreChain(String userName, Object credential, boolean domainProvided,
                                            List<String> userStorePreferenceOrder) throws UserStoreException {
 
+        // If domain name is provided, directly authenticate using the corresponding user store.
+        if (domainProvided) {
+            String providedDomainName = this.getUserStore(userName).getDomainName();
+            // Check whether provided domain is in the preference list.
+            if (!userStorePreferenceOrder.contains(providedDomainName)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Authentication failure. Wrong username or password is provided.");
+                }
+                handleOnAuthenticateFailure(ErrorMessages.ERROR_CODE_ERROR_WHILE_AUTHENTICATION.getCode(),
+                        String.format(ErrorMessages.ERROR_CODE_ERROR_WHILE_AUTHENTICATION.getMessage(),
+                                "Authentication failed"), userName, credential);
+                throw new UserStoreException("Authentication failed. Invalid username or password.");
+            }
+            UserStoreManager userStoreManager = this.getSecondaryUserStoreManager(providedDomainName);
+            if (!(userStoreManager instanceof AbstractUserStoreManager)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("UserStoreManager is not an instance of AbstractUserStoreManager hence authenticate " +
```

Comment:
Shall we end the sentence from `instance of AbstractUserStoreManager` and begin a new sentence? WDYT?

Reply:
And shall we do the same in other relevant places?

Reply:
Done